### ISSUE TITLE
instances and classes skulpt nuances 

### ISSUE THREAD

**Issue by s-cork** [skulpt/skulpt#1075]
This is something that has been on my mind when working with skulpt and some things I came across when working on #1040 

currently the implementation for classes and instances with the `genericGetattr` has some quirks.

```python
>>> class A:
...    pass
...
>>> a = A()
>>> A.__class__
<class '__main__.A'>  # should be <class 'type'>
>>> type(A)
<class 'type'>
>>> a.__class__
<class '__main__.A'>
```

with builtins
```python
>>> x = tuple()
>>> x.__class__
AttributeError: 'tuple' object has no attribute '__class__' # should be <class 'tuple'>
>>> tuple.__class__
AttributeError: 'tuple' object has no attribute '__class__' # should be <class 'type'>
```
despite the code for tuple doing
```javascript
this.__class__ = Sk.builtin.tuple;
```

Similarly the way that Skulpt handles __bases__, __name__ and __mro__ (which shouldn't be accessed by instances of classes) currently can be
```python
>>> a.__name__
A # should be attribute error
>>> a.__bases__
(<class 'object'>,)  # should be attribute error
>>> a.__mro__
(<class '__main__.A'>, <class 'object'>) # should be attribute error
>>> x = int()
>>> int.__name__
'int'
>>> x.__name__
'int' # should be attribute error
```
---

**Comment by s-cork** [comment]
Have thought about this a little bit today

_nb: I haven't considered backwards compatibility to python 2_

if we look at `type`:
```python
>>> for key in type.__dict__.keys():
...   print(key, type(type.__dict__[key]))
... 
```

<details>

<summary>

</summary>

```python
__repr__ <class 'wrapper_descriptor'>
__call__ <class 'wrapper_descriptor'>
__getattribute__ <class 'wrapper_descriptor'>
__setattr__ <class 'wrapper_descriptor'>
__delattr__ <class 'wrapper_descriptor'>
__init__ <class 'wrapper_descriptor'>
__new__ <class 'builtin_function_or_method'>
mro <class 'method_descriptor'>
__subclasses__ <class 'method_descriptor'>
__prepare__ <class 'classmethod_descriptor'>
__instancecheck__ <class 'method_descriptor'>
__subclasscheck__ <class 'method_descriptor'>

__dir__ <class 'method_descriptor'>

__sizeof__ <class 'method_descriptor'>
__basicsize__ <class 'member_descriptor'>
__itemsize__ <class 'member_descriptor'>
__flags__ <class 'member_descriptor'>
__weakrefoffset__ <class 'member_descriptor'>
__base__ <class 'member_descriptor'>
__dictoffset__ <class 'member_descriptor'>

__mro__ <class 'member_descriptor'>

__name__ <class 'getset_descriptor'>

__qualname__ <class 'getset_descriptor'>

__bases__ <class 'getset_descriptor'>

__module__ <class 'getset_descriptor'>
__abstractmethods__ <class 'getset_descriptor'>

__dict__ <class 'getset_descriptor'>

__doc__ <class 'getset_descriptor'>
__text_signature__ <class 'getset_descriptor'>
```

</details>


and compare to the same with `object`


```python
>>> for key in object.__dict__.keys():
...   print(key, type(object.__dict__[key]))
... 
```
<details>

<summary>

</summary>

```python
__repr__ <class 'wrapper_descriptor'>
__hash__ <class 'wrapper_descriptor'>
__str__ <class 'wrapper_descriptor'>
__getattribute__ <class 'wrapper_descriptor'>
__setattr__ <class 'wrapper_descriptor'>
__delattr__ <class 'wrapper_descriptor'>
__lt__ <class 'wrapper_descriptor'>
__le__ <class 'wrapper_descriptor'>
__eq__ <class 'wrapper_descriptor'>
__ne__ <class 'wrapper_descriptor'>
__gt__ <class 'wrapper_descriptor'>
__ge__ <class 'wrapper_descriptor'>
__init__ <class 'wrapper_descriptor'>
__new__ <class 'builtin_function_or_method'>
__reduce_ex__ <class 'method_descriptor'>
__reduce__ <class 'method_descriptor'>
__subclasshook__ <class 'classmethod_descriptor'>
__init_subclass__ <class 'classmethod_descriptor'>
__format__ <class 'method_descriptor'>
__sizeof__ <class 'method_descriptor'>

__dir__ <class 'method_descriptor'>

__class__ <class 'getset_descriptor'>

__doc__ <class 'str'>
```

</details>

And for just a class
```python
>>> class A:
...   pass
...
>>> for key in A.__dict__.keys():
...   print(key, type(A.__dict__[key]))
... 

```
<details>
<summary></summary>

```python

__module__ <class 'str'>

__dict__ <class 'getset_descriptor'>

__weakref__ <class 'getset_descriptor'>
__doc__ <class 'NoneType'>
```

</details>

Does this suggest where certain attributes should live?


## Examples
**Example1** 
At the moment all classes and instances of those class get an attribute `"__class__"`
But actually only object should have an attribute `"__class__"` which is of type `getset_descriptor` and would return `ob$type`

**Example 2**
instances of type have a `$d` dictionary. Typically in the dictionary goes `__mro__` and `__bases__`. with `__name__` appearing in the prototype of a class. 

But actually only type should have an attribute `"__name__"`, `"__mro__"` and `"__bases__"` which would return `tp$name`, `tp$mro` and `tp$bases`.

Since instances of `object` are not instances of `type` then they would not have access to such.

**Example 3**
Most classes don't need a `$d` property. Only their instances.

**Comment by s-cork** [comment]
Have done some more work on this and this is where I'm at with it...

idea:
- implement a `MappingProxy` type in skulpt
- implement a `GetSetDescriptor` type in skulpt
- `classes` have `"$d"` as a `MappingProxy` type
- instances remain unchanged and have `"$d"` as a dictionary

---

`MappingProxy` type would be like a dictionary in that it had certain methods like `mp$lookup` but it would be much faster because keys would only be javascript strings.

---

`GetSetDescriptor` would be similar to pypy/Cpython implementation. In that you can call it with `fget`, `fset`, `fdel` and these would be assigned to the slots` tp$descr_get`, `tp$descr_set`, `tp$descr_del`. 

---

Rationale - gives the ability to fully implement `__dict__` and only check whether something is in the `["$d"]` property of instances and classes. 

`typeLookup` would be fast with the `MappingProxy` type

---
<details>
<summary>
worked example
</summary>

```javascript
// pseudo code
object["$d"] = Sk.builtin.mappingProxy(
    {"__class__" : new Sk.builtin.getsetDescriptor(
  fget=function(obj) {return obj.ob$type}, 
  fset = //some function where we raise TypeError if not a class)
}
)
```
so "__class__" does not need to be defined in `klass` anymore.

```python

>>> class A:
...    pass
...
>>> a = A()
>>> a.__class__
# do I have __class__ in my "$d"?       No
# do type lookup:
# does A have __class__ in its "$d"?   No
# does object have __class__ in it's "$d"? yes
# does it have tp$descr_get method? yes
# return tp$descr_get.call 
<class '__main__.A'>
>>> a.__class__ = 'foo'
# as before but since tp$desc_set raises a TypeError
TypeError
>>> A.__dict__
mappingproxy({'__module__': '__main__', 
            '__dict__': <attribute '__dict__' of 'A' objects>,
            '__doc__': None})
```

</details>

---
<details>
<summary>
Further thoughts on `type.prototype.tp$getattr`
</summary>


Currently `type` has no `tp$mro` attribute which should be 

```python
>>> type.__mro__
(<type 'type'>, <type 'object'>)
```

this means that the `type.prototype.tp$getattr` algorithm can be adjusted to:


```javascript
Sk.builtin.type.prototype.tp$getattr = function (pyName, canSuspend) {
    var tp = this.ob$type;
    var descr;
    descr = Sk.builtin.type.typeLookup(this, pyName) || Sk.builtin.type.typeLookup(tp, pyName);
    // then call the tp$descr_get or return descr
```
</details>

---

I've got proof of concept on many of the above points. But it's far from done!

**Comment by s-cork** [comment]
update... I've been working on this over the past couple of days during lockdown. 

I have implemented a `mappingproxy` and a `getset_descriptor` type.

As a benchmark to their utility, I removed all references to the word `__class__` from the codebase and put a `getset_descriptor` in the `mappingproxy` of `object` (as per cpython). 

```python
>>> object.__dict__["__class__"]
<attribute '__class__' of 'object' objects>
>>> type(object.__dict__["__class__"])
<class 'getset_descriptor'>
```

This way a call to `__class__` works up `tp$getattr` until it hits the `object`'s `mappingproxy` at which point it calls the descriptor and boom, what's my `ob$type`?
I've got it passing all tests!

---

next phases:
- [ ] remove all references to `__name__`, `__bases__` and `__mro__` apart from in `type`'s `mappingproxy`
- [ ] create a `mappingproxy` for all `builtins` which includes their dunder methods.
- [ ] change the way `tp$getattr` works. Go through the `dict` of instances of the `mappingproxy`. 
---
